File: 0981-time-based-key-value-store/0981-time-based-key-value-store.py
from sortedcontainers import SortedDict
class TimeMap:

    def __init__(self):
        self.m = {}

    def set(self, key: str, value: str, timestamp: int) -> None:
        if key not in self.m:
            self.m[key] = SortedDict()
        self.m[key][timestamp] = value

    def get(self, key: str, timestamp: int) -> str:
        if key not in self.m:
            return ''
        it = self.m[key].bisect_right(timestamp)
        if it == 0:
            return ''
        return self.m[key].peekitem(it-1)[1]


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)

File: 0981-time-based-key-value-store/0981-time-based-key-value-store.cpp
class TimeMap {
public:
    unordered_map<string, map<int,string>> m;
    TimeMap() {
        
    }
    
    void set(string key, string value, int timestamp) {
        m[key][timestamp] = value;
    }
    
    string get(string key, int timestamp) {
        if (m.find(key) == m.end())
            return "";
        auto& a = m[key];
        auto it = a.lower_bound(timestamp);
        if (it == a.end())
            return a.rbegin()->second;
        if (it->first == timestamp)
            return it->second;
        if (it == a.begin())
            return "";
        return (--it)->second;
    }
};

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap* obj = new TimeMap();
 * obj->set(key,value,timestamp);
 * string param_2 = obj->get(key,timestamp);
 */